pkg/api/v1beta1/types.go

```diff
@@ -307,6 +307,19 @@ type TypeMeta struct {
 	APIVersion        string    `json:"apiVersion,omitempty" description:"version of the schema the object should have"`
 	Namespace         string    `json:"namespace,omitempty" description:"namespace to which the object belongs; must be a DNS_SUBDOMAIN; 'default' by default"`
 
+	// GenerateName indicates that the name should be made unique by the server prior to persisting
+	// it. A non-empty value for the field indicates the name will be made unique (and the name
+	// returned to the client will be different than the name passed). The value of this field will
+	// be combined with a unique suffix on the server, and takes precedence over the Name field.
+	// The provided value must be valid within the rules for Name, and may be truncated by the length
+	// of the suffix required to make the value unique on the server.
+	//
+	// If this field is specified, the server will NOT return a 409 if Name exists - instead, it will
+	// either return 201 Created or 500 with Reason TryAgainLater indicating a unique name could not
+	// be found in the time allotted, and the client should retry (optionally after the time indicated
+	// in the Retry-After header).
+	GenerateName string `json:"generateName,omitempty" description:"an optional prefix to use to generate a unique name; has the same validation rules as name; optional and overrides name if specified"`
```

Comment:
This was another GET -> POST gut feel.  If I want to copy a bunch of things from namespace 1 to 2 that had generateName set (like a replication controller, or a service), do I also want to generate a name in the new namespace?  The argument for is that you'd be able to GET -> POST in the same namespace to copy something.  The argument against is that any interlinks would be broken by name generation (if you exported multiple things).

Which way would you land?

----- Original Message -----
> > @@ -307,6 +307,19 @@ type TypeMeta struct {
> >  	APIVersion        string    `json:"apiVersion,omitempty"
> >  	description:"version of the schema the object should have"`
> >  	Namespace         string    `json:"namespace,omitempty"
> >  	description:"namespace to which the object belongs; must be a
> >  	DNS_SUBDOMAIN; 'default' by default"`
> >  
> > +	// GenerateName indicates that the name should be made unique by the
> > server prior to persisting
> > +	// it. A non-empty value for the field indicates the name will be made
> > unique (and the name
> > +	// returned to the client will be different than the name passed). The
> > value of this field will
> > +	// be combined with a unique suffix on the server, and takes precedence
> > over the Name field.
> > +	// The provided value must be valid within the rules for Name, and may be
> > truncated by the length
> > +	// of the suffix required to make the value unique on the server.
> > +	//
> > +	// If this field is specified, the server will NOT return a 409 if Name
> > exists - instead, it will
> > +	// either return 201 Created or 500 with Reason TryAgainLater indicating
> > a unique name could not
> > +	// be found in the time allotted, and the client should retry (optionally
> > after the time indicated
> > +	// in the Retry-After header).
> > +	GenerateName string `json:"generateName,omitempty" description:"an
> > optional prefix to use to generate a unique name; has the same validation
> > rules as name; optional and overrides name if specified"`
> 
> Do we really want this to override name? I think this should behave like
> regular defaulting: a name is generated only if not provided.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/3789/files#r23820525